apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/Input.kt

```diff
@@ -1,64 +1,27 @@
 package com.apollographql.apollo3.api
 
-import kotlin.jvm.JvmField
-import kotlin.jvm.JvmStatic
-
 /**
- * A wrapper class of provide [value] with a [defined] state.
+ * [Input] is a holder class that has a value only if isPresent is true
  *
- * Wrapper can be in one of 2 states:
- * - *defined* (`defined == true`) means reference to [value] is set explicitly
- * - *undefined* (`defined == false`) means reference is not set.
+ * It uses the same trick as [kotlin.Result] to store the value as Any so that Result(Absent) can be any type of Result<T>
  *
- * It provides a convenience way to distinguish the case when [value] is provided explicitly and should be
- * serialized (even if it's null) and the case when [value] is undefined (means it won't be serialized).
+ * It provides a convenient way to distinguish the case when a value is provided explicitly and should be
+ * serialized (even if it's null) and the case where it's absent and shouldn't be serialized.
  */
-class Input<V> internal constructor(
-    @JvmField val value: V?,
-    @JvmField val defined: Boolean
-) {
-
-  override fun equals(other: Any?): Boolean {
-    if (this === other) return true
-    if (other !is Input<*>) return false
-
-    if (value != other.value) return false
-    if (defined != other.defined) return false
+class Input<out V>(private val value: Any?) {
```

Comment:
Should we make it more idiomatic and represent it as sealed class?

Reply:
Yup, I took inspiration from the kotlin [Result](https://github.com/JetBrains/kotlin/blob/master/libraries/stdlib/src/kotlin/util/Result.kt) class but sealed class make better use of type inference, etc.. I just pushed the change

Reply:
I just bumped into that: https://github.com/Kotlin/KEEP/blob/1827a4d3da410f70797522a156405dcbd3b1069a/proposals/stdlib/result.md#representing-as-a-sealed-class

The reason `Result` is not a sealed class in the stdlib is performance since they want the values to be stored unboxed. I don't think that changes anything in our case since this is not going to be used in any "hot" path, it's fine to pay the boxing price I think.